File: 프로그래머스/2/131704. 택배상자/택배상자.cs
using System;
using System.Collections.Generic;

public class Solution {
    public int solution(int[] order) {
        int answer = 0;
        var stack = new Stack<int>();
        int n = order.Length;
        int cur = 1;
        
        foreach(int target in order)
        {
            while(cur <= n && (stack.Count == 0 || stack.Peek() != target))
            {
                stack.Push(cur++);
            }
            if(stack.Count > 0 && stack.Peek() == target)
            {
                stack.Pop();
                answer++;
            }
            else break;
        }
        return answer;
    }
}

File: 프로그래머스/2/131704. 택배상자/택배상자.cpp
#include <string>
#include <vector>
#include <bits/stdc++.h>

using namespace std;

int solution(vector<int> order) {
    int answer = 0;
    queue<int> q;
    stack<int> stk;
    
    for(int i = 1; i <= order.size(); i++)
    {
        q.push(i);
    }
    
    for(int i = 0; i < order.size(); i++)
    {
        if(stk.size() && stk.top() == order[i])
        {
            stk.pop();
            answer++;
            continue;
        }
        while(q.size() && q.front() != order[i])
        {
            stk.push(q.front());
            q.pop();
        }
        
        if(q.size() && q.front() == order[i])
        {
             q.pop();
             answer++;
        }
        else
            break;            
    }
    return answer;
}